### PR TITLE
Use new compiler symbol search helpers that are optimized for string names searches.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/26325 (merged) and https://github.com/dotnet/roslyn/pull/26330.  in this PR i've updated the IDE to forward certain helpers to these more efficient implementations.  

This helps things out by more quickly being able to determine if a type even contains a member with name, and thus whether or not it should even be hydrated into a symbol and have its members created.  Previous we would have to do a linear scan on all the members in a type to determine this.  Now this data is in a set which can be queried much more efficiently.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @jcouv @dotnet/roslyn-ide .  Note: this doesn't need to reviewed until dependent PRs go through.

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler for new APIs exposed from the Compilation objects.   Tagging @dotnet/roslyn-ide for usages of that new API.

**Comment by CyrusNajmabadi** [comment]
Tagging @jcouv for being an honorary @dotnet/roslyn-compiler  and @dotnet/roslyn-ide dev :D

File: src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder.cs [hunk 1]
```diff
@@ -39,15 +40,30 @@ internal static partial class DeclarationFinder
...
                 var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
-                var symbolsWithName = compilation.GetSymbolsWithName(query.GetPredicate(), filter, cancellationToken)
-                                                 .Select(s => new SymbolAndProjectId(s, project.Id))
-                                                 .ToImmutableArray();
+
+                var symbols = isExactNameSearch
+                    ? compilation.GetSymbolsWithName(query.Name, filter, cancellationToken)
+                    : compilation.GetSymbolsWithName(query.GetPredicate(), filter, cancellationToken);
+
+                var symbolsWithName = symbols.Select(s => new SymbolAndProjectId(s, project.Id)) .ToImmutableArray();
```
Comment (Neme12) [inline src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder.cs]: SelectAsArray?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder.cs]: Will do.

File: src/Workspaces/Core/Portable/Workspace/Solution/Project.cs [hunk 2]
```diff
@@ -224,6 +224,12 @@ internal TextDocumentState GetAdditionalDocumentState(DocumentId documentId)
             return _projectState.GetAdditionalDocumentState(documentId);
         }
 
+        internal async Task<bool> ContainsSymbolsWithNameAsync(string name, SymbolFilter filter, CancellationToken cancellationToken)
+        {
+            return this.SupportsCompilation &&
+                   await _solution.State.ContainsSymbolsWithNameAsync(Id, name, filter, cancellationToken).ConfigureAwait(false);
+        }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/Project.cs]: Near copy of the function below

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs [hunk 3]
```diff
@@ -1780,6 +1780,25 @@ public Task<MetadataReference> GetMetadataReferenceAsync(ProjectReference projec
...
+            // it looks like declaration compilation doesn't exist yet. we have to build full compilation
+            var compilation = await GetCompilationAsync(id, cancellationToken).ConfigureAwait(false);
+            if (compilation == null)
+            {
+                // some projects don't support compilations (e.g., TypeScript) so there's nothing to check
+                return false;
+            }
+
+            return compilation.ContainsSymbolsWithName(name, filter, cancellationToken);
+        }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs]: Near copy of the function below

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 4]
```diff
@@ -3050,12 +3050,13 @@ public override IEnumerable<ISymbol> GetSymbolsWithName(Func<string, bool> predi
             return new PredicateSymbolSearcher(this, filter, predicate, cancellationToken).GetSymbolsWithName();
         }
 
+#pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: these overloads shouldnot be a problem.  There is no conflict/ambiguity between them and the existing overloads that take delegates.

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler for public API additions to Compilation.
Tagging @dotnet/roslyn-ide for the updates to the IDE code.

My previous PRs https://github.com/dotnet/roslyn/pull/26325 and https://github.com/dotnet/roslyn/pull/26330 improved the internal behavior of ContainsSymbolByName and GetSymbolsByName.  Both by making it so we almost never need to realize symbols unless we have a match, and so that name lookups are not linear.   Previously, we would realize *all* top level symbols including namespace, types, **and** members.  And then we'd linearly walk all of those for a match.  this was massively expensive and produced tons of garbage by forcing so many symbols to be made.  Now, we can sit directly on the decl-table, which already has these indices.

This PR now updates the IDE to use the faster compiler entrypoints which can be much smarter by using our actual decl tables.  

From https://github.com/dotnet/roslyn/pull/26330, measurements show around a 12x improvement over calling the new API over hte old one when the name is known ahead of time.

**Comment by CyrusNajmabadi** [comment]
Tagging @jcouv

**Comment by CyrusNajmabadi** [comment]
test windows_debug_spanish_unit32_prtest

**Comment by CyrusNajmabadi** [comment]
test ubuntu_16_debug_prtest

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler for public API additions to Compilation.
Tagging @dotnet/roslyn-ide for the updates to the IDE code.

Can you please take a look at this helpful change that speeds up IDE searches substantively.  Thanks!

@agocke This is just the utilization of the APIs you already approved.  Can you take a look?

**Comment by CyrusNajmabadi** [comment]
@jinujoseph can this get an IDE buddy?

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-compiler for public API additions to Compilation.
Tagging @dotnet/roslyn-ide for the updates to the IDE code.

Can you please take a look at this helpful change that speeds up IDE searches substantively. Thanks!

@agocke This is just the utilization of the APIs you already approved. Can you take a look?

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 5]
```diff
@@ -31,6 +31,8 @@ Microsoft.CodeAnalysis.Operations.ITupleBinaryOperation.OperatorKind.get -> Micr
 Microsoft.CodeAnalysis.Operations.ITupleBinaryOperation.RightOperand.get -> Microsoft.CodeAnalysis.IOperation
 Microsoft.CodeAnalysis.Operations.ITupleOperation.NaturalType.get -> Microsoft.CodeAnalysis.ITypeSymbol
 Microsoft.CodeAnalysis.Platform.Arm64 = 6 -> Microsoft.CodeAnalysis.Platform
+abstract Microsoft.CodeAnalysis.Compilation.ContainsSymbolsWithName(string name, Microsoft.CodeAnalysis.SymbolFilter filter = Microsoft.CodeAnalysis.SymbolFilter.TypeAndMember, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> bool
```
Comment (jcouv) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt]: Tagging @gafter @jaredpar for public API change

File: src/Compilers/Core/Portable/Compilation/Compilation.cs [hunk 6]
```diff
@@ -2907,19 +2907,21 @@ internal string GetMessage(ITypeSymbol source, ITypeSymbol destination)
...
         public abstract IEnumerable<ISymbol> GetSymbolsWithName(Func<string, bool> predicate, SymbolFilter filter = SymbolFilter.TypeAndMember, CancellationToken cancellationToken = default(CancellationToken));
 
+#pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
         /// <summary>
         /// Return true if there is a source declaration symbol name that matches the provided name.
         /// This may be faster than <see cref="ContainsSymbolsWithName(Func{string, bool}, SymbolFilter, CancellationToken)"/>
         /// when predicate is just a simple string check.
         /// </summary>
-        internal abstract bool ContainsSymbolsWithName(string name, SymbolFilter filter = SymbolFilter.TypeAndMember, CancellationToken cancellationToken = default(CancellationToken));
+        public abstract bool ContainsSymbolsWithName(string name, SymbolFilter filter = SymbolFilter.TypeAndMember, CancellationToken cancellationToken = default(CancellationToken));
```
Comment (jcouv) [inline src/Compilers/Core/Portable/Compilation/Compilation.cs]: :memo: Notes for reviewers, the tests for these APIs were added back in [https://github.com/dotnet/roslyn/pull/26325](https://github.com/dotnet/roslyn/pull/26325)

File: src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder.cs [hunk 7]
```diff
@@ -39,15 +41,36 @@ internal static partial class DeclarationFinder
 
             using (Logger.LogBlock(FunctionId.SymbolFinder_Project_AddDeclarationsAsync, cancellationToken))
             {
-                if (!await project.ContainsSymbolsWithNameAsync(query.GetPredicate(), filter, cancellationToken).ConfigureAwait(false))
+                var syntaxFacts = project.LanguageServices.GetService<ISyntaxFactsService>();
+
+                // If this is an exact query, we can speed things up by just calling into the
+                // compilation entrypoints that take a string directly.
+                //
+                // the search is 'exact' it's either an exact-case-sensitive search,
```
Comment (jcouv) [inline src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder.cs]: missing word `if`?

Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/Declarations/DeclarationFinder.cs]: Fixed!

**Comment by CyrusNajmabadi** [comment]
@gafter @jaredpar Can you look at this small public API addition?  Thanks!

**Comment by jcouv** [comment]
@gafter for review of compiler change. Thanks

**Comment by gafter** [comment]
> @gafter for review of compiler change. Thanks

@jcouv The principal compiler change is a public API addition. I've scheduled a review of that API for tomorrow at 2pm. I'll comment here on the result of that review.

**Comment by gafter** [comment]
@CyrusNajmabadi 
Reviewed this just now. One small change is requested: please document the language-specific behavior of `name` in `Compilation.ContainsSymbolsWithName` and `Compilation.GetSymbolsWithName` in the new APIs, and in the existing related APIs. e.g. mention that it is case-sensitive for C# and case-insensitive for VB if that is the case.

**Comment by gafter** [comment]
@CyrusNajmabadi 
Also please bring up-to-date with master.
Please confirm if it would be OK with you if we "Squash and merge" to complete this PR.

**Comment by CyrusNajmabadi** [comment]
> Reviewed this just now. One small change is requested: please document the language-specific behavior of name in Compilation.ContainsSymbolsWithName and Compilation.GetSymbolsWithName in the new APIs, and in the existing related APIs. e.g. mention that it is case-sensitive for C# and case-insensitive for VB if that is the case.

Absolutely.

> Also please bring up-to-date with master.

Will do!

> Please confirm if it would be OK with you if we "Squash and merge" to complete this PR.

I'm 100% ok with that.

**Comment by gafter** [comment]
Pending merge approval.

**Comment by jaredpar** [comment]
approved

**Comment by jcouv** [comment]
@jinujoseph Can you also approve for 15.8 ask-mode since there is an IDE change? Thanks

**Comment by jinujoseph** [comment]
Approved to merge for 15.8.Preview3

**Comment by jcouv** [comment]
@gafter This PR is ready (approved by Jared and Jinu). But I don't know what "Squash and merge" is safe on a PR that includes merge commits from master...

**Comment by CyrusNajmabadi** [comment]
@jcouv I'm happy to rebase and submit for you to merge.  Would you prefer that?

**Comment by gafter** [comment]
@CyrusNajmabadi THANK YOU!!!!!

**Comment by CyrusNajmabadi** [comment]
Thank you!

**Pull request merged**